### PR TITLE
Add code

### PR DESCRIPTION
**Pull request by brownjohnf**

### CODE REVIEW

**Comment by brownjohnf** [comment]
@hedss This is a refactor/rewrite of the last version you saw. It handles a lot more edge cases, and I started playing around with classes, which cleans up/internalizes a lot of the logic. I'm now going to work on making it a module, so it can be installed via npm instead of a clone/docker.

Once that's set up, I'll work on getting it into a procbot, but this is basically the code I've been using locally to do the k8s stuff I've been working on.

File: src/lib/cli/keyfctl.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      res.tap(frames => {
+        console.log('==> All frames')
+        for (let frame of frames) {
+          // Output some feedback about the frame history
+          console.log(utils.printFormatFrame(frame))
+        }
+      })
+    }
+
+    res = res.filter(frame => {
```
Comment (hedss) [inline src/lib/cli/keyfctl.js:96]: I see what you're doing here with augmenting the promise, but I've never seen it written in this kind of style before. I can't decide if it's disgusting or utterly genius. :)

File: src/lib/shared/configmap.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  utils = require('../shared/utils')
+
+module.exports = class Configmap {
+  constructor(revision, componentName) {
+    this.revision  = revision
+    this.errors    = []
+    this.rationale = []
+    this.valid     = true
+
+    return file.readAt(
```
Comment (hedss) [inline src/lib/shared/configmap.js]: It's a pretty bad anti-pattern to return anything (esp. something that mutates the return value of 'new') from a constructor. If you want to start something when you construct the object, you could do something like:
```
constructor(args) {
    ....
    this.initPromise = doSomething();
}

callAfterConstruction() {
    return this.initPromise;
}
```
That way you cleanly return from the constructor ASAP as intended, but can then call a method that will either have resolved or will return once the task started in the constructor has resolved.

File: src/lib/shared/deployment.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+module.exports = class Deployment {
+  constructor(revision, timestamp, component) {
+    this.component = component
+    this.revision = revision
+    this.timestamp = timestamp
+    this.errors = []
+    this.rationale = []
+    this.valid = true
+
+    return file.readAt(
```
Comment (hedss) [inline src/lib/shared/deployment.js]: Constructor anti-pattern again.

**Review by hedss — COMMENTED** [review]
A few comments off the top of my head, will look in more detail next week when I'm not as tired!

**Pull request merged**